### PR TITLE
Add config for tombstone record with a default behavior of ingesting …

### PR DESCRIPTION
**Pull request by sfc-gh-japatel**
…empty json

https://snowflakecomputing.atlassian.net/browse/SNOW-298877

Background:
- Tombstone records: Records with null value. 
- Clients can use community converters with Tombstone Handler SMT (Single Message Transformation) - https://docs.confluent.io/platform/current/connect/transforms/tombstonehandler.html
- But, if they use converters defined by our KC, SMT is not supported. 
- Current behavior of TS records are that they are representing an empty json string in Snowflake Table. 
- To avoid doing this and prevent us from doing a breaking change, we can fallback to default behavior and introduce another behavior which would ignore those null values. 
- Remember, Snowflake is meant to be append only and this tombstone records are not used for removing corresponding keys from Snowflake Table. (Original idea of TS record is to be able to remove those keys from destination table if values are null)
- Custom converters create an empty Json String(Check default ctor for SnowflakeRecordContent). So, I introduced a flag(`isNullValueRecord`) for Sink service to identify if converter received an null value which still resulted into a non null content (representing an empty json string)

### Testing
- I had long ago added a test to verify that tombstone records are resulting into an empty string. 
- The above test had default behavior of "DEFAULT" which gave the same result. 
- Added one more test with different config "IGNORE" which proved the buffer to be empty and nothing on stage as well. 
- Here is the connector config: https://jsonblob.com/1c34495a-c88b-11eb-ba15-1ba7dd08f3a3
  - If debug logs are enabled, here is what it looks like: 
  - ``` Null valued record from topic 'SnowflakeSink_JP2', partition 0 and offset 2 was skipped. (com.snowflake.kafka.connector.internal.SnowflakeSinkServiceV1:75)```

Required Reviewers:
@sfc-gh-rramachandran and @sfc-gh-zli (As he would have some more context if I missed an edge case)

### CODE REVIEW

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java [hunk 1]
```diff
@@ -87,6 +99,77 @@ public void insert(final Collection<SinkRecord> records) {
...
+            record.topic(),
+            record.kafkaPartition(),
+            record.kafkaOffset());
+        isRecordValueNull = true;
+      }
+    }
+
+    if (isRecordValueNull) {
+      if (behaviorOnNullValues
+          .toString()
```
Comment (sfc-gh-rramachandran) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:154]: Aren't the 2 enums the same? Why can't we do a simple equality check

File: src/main/java/com/snowflake/kafka/connector/records/SnowflakeRecordContent.java [hunk 2]
```diff
@@ -109,4 +124,20 @@ int getSchemaID() {
...
+   * Check if primary reason for this record content's value to be an empty json String, a null
+   * value?
+   *
+   * <p>i.e if value passed in by record is empty json node (`{}`), we don't interpret this as null
+   * value.
+   *
+   * @return true if content value is empty json node as well as isNullValueRecord is set to true.
+   */
+  public boolean isRecordContentValueNull() {
+    if (content != null && content[0].isEmpty() && isNullValueRecord) {
```
Comment (sfc-gh-rramachandran) [inline src/main/java/com/snowflake/kafka/connector/records/SnowflakeRecordContent.java:138]: Under what scenario is content null? Is it added just for safety?
Also for my understanding, does a null record json look like this
`[
    null
]`

**Review by sfc-gh-rramachandran — COMMENTED** [review]
Left some comments

File: src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java [hunk 3]
```diff
@@ -416,4 +432,53 @@ public static KafkaProvider of(final String kafkaProviderStr) {
...
+
+  /* The allowed values for tombstone records. */
+  public enum BehaviorOnNullValues {
+    // Ignore would filter out records which has null value, but a valid key.
+    IGNORE,
+
+    // Default as the name suggests, would be a default behavior which will not filter null values.
+    // This will put an empty JSON string in corresponding snowflake table.
+    // Using this means we will fall back to old behavior before introducing this config.
+    DEFAULT,
```
[the remark this answers is not in the document]
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java:444]: Ahh yes, will do that.

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java [hunk 4]
```diff
@@ -87,6 +99,77 @@ public void insert(final Collection<SinkRecord> records) {
...
+   *
+   * <p>If the value is an empty JSON node, we could assume the value passed was null.
+   *
+   * @see com.snowflake.kafka.connector.records.SnowflakeJsonConverter#toConnectData when bytes ==
+   *     null case
+   * @param record record sent from Kafka to KC
+   * @return true if we would skip adding it to buffer -> skip to internal stage and hence skipped
+   *     inside SF Table
+   */
+  private boolean maybeSkipOnNullValue(SinkRecord record) {
```
[the remark this answers is not in the document]
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:118]: How about `shouldSkipNullValue()`?

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java (hunk 1 above)
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:154]: Yes, no need for equalsIgnoreCase, I copy pasted above code. Enum equality (==) should also work.

File: src/main/java/com/snowflake/kafka/connector/records/SnowflakeRecordContent.java (hunk 2 above)
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/records/SnowflakeRecordContent.java:138]: I added content != null and content[0].isEmpty for sanity check. 
A null record looks like this `{}`
(Check constructor for SnowflakeRecordContent)
```
public SnowflakeRecordContent() {
    content = new JsonNode[1];
    content[0] = MAPPER.createObjectNode();
    brokenData = null;
  }

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (sfc-gh-rramachandran) [reply src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:118]: Sounds good!

File: src/main/java/com/snowflake/kafka/connector/records/SnowflakeRecordContent.java (hunk 2 above)
Comment (sfc-gh-rramachandran) [reply src/main/java/com/snowflake/kafka/connector/records/SnowflakeRecordContent.java:138]: Hmm, but isn't content an array? So does it look like this then?
`[ {} ]`

Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/records/SnowflakeRecordContent.java:138]: That is correct. Looks like we go through the array and create the content. (Meta and actual content)
https://github.com/snowflakedb/snowflake-kafka-connector/blob/master/src/main/java/com/snowflake/kafka/connector/records/RecordService.java#L129-L139

Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/records/SnowflakeRecordContent.java:138]: So ultimately this is what it looks like: 
{"content":{},"meta":{"topic":"test","offset":0,"partition":0,"schema_id":0,"key":"test"}}

File: src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java [hunk 5]
```diff
@@ -330,7 +334,19 @@ static ConfigDef newConfigDef() {
...
-            "Whether kafka is running on Confluent code, self hosted or other managed service");
+            "Whether kafka is running on Confluent code, self hosted or other managed service")
+        .define(
+            BEHAVIOR_ON_NULL_VALUES_CONFIG,
+            Type.STRING,
+            BehaviorOnNullValues.DEFAULT.toString(),
+            BehaviorOnNullValues.VALIDATOR,
+            Importance.LOW,
+            "How to handle records with a null value (i.e. Kafka tombstone records)."
+                + " Valid options are 'DEFAULT' and 'IGNORE'.",
```
[the remark this answers is not in the document]
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java:345]: There can be another option which could fail the connector, hence keeping binary is not a good idea. Also, I have kept this consistent naming as it is the same in S3SinkConnector.

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java [hunk 6]
```diff
@@ -87,6 +99,75 @@ public void insert(final Collection<SinkRecord> records) {
...
+              record.kafkaPartition(),
+              record.kafkaOffset());
+          isRecordValueNull = true;
+        }
+      }
+    } else {
+      // Else, it is one of the community converters.
+      // Tombstone handler SMT can be used but we need to check here if value is null if SMT is not
+      // used
+      if (record.value() == null) {
```
[the remark this answers is not in the document]
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:142]: Umm, no.. Record.key are present for tombstone records since they help identify what key needs to be removed from the corresponding table. (See Tombstone records: https://rmoff.net/2020/11/03/kafka-connect-ksqldb-and-kafka-tombstone-messages/)
(Since we dont have concept of upserts.) Our Connector is append only.

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java [hunk 7]
```diff
@@ -87,6 +99,75 @@ public void insert(final Collection<SinkRecord> records) {
...
+            "Record value is null for topic {}, partition {} and offset {}",
+            record.topic(),
+            record.kafkaPartition(),
+            record.kafkaOffset());
+        isRecordValueNull = true;
+      }
+    }
+
+    if (isRecordValueNull) {
+      if (behaviorOnNullValues == SnowflakeSinkConnectorConfig.BehaviorOnNullValues.IGNORE) {
```
[the remark this answers is not in the document]
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:153]: Yes, makes sense to do it at the beginning. Let me do it.

File: src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java:345]: https://docs.confluent.io/kafka-connect-s3-sink/current/configuration_options.html

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:153]: If we do the early check, we will miss the ability to print that they have default value and we are sending in an empty string. But, I agree we will save a few cpu cycles.

**Review by sfc-gh-zli — APPROVED** [review]
